indexers/keyvalue/LevelDBIndexer/tests/test_leveldbindexer.py

```diff
@@ -1,61 +1,195 @@
 
 from .. import LevelDBIndexer
 
-cur_dir = os.path.dirname(os.path.abspath(__file__))
-
-
-def rm_files(file_paths):
-    for file_path in file_paths:
-        if os.path.exists(file_path):
-            if os.path.isfile(file_path):
-                os.remove(file_path)
-            elif os.path.isdir(file_path):
-                shutil.rmtree(file_path, ignore_errors=False, onerror=None)
-
-
-def run_test(indexer):
-    def create_document(doc_id, text, weight, length):
-        d = jina_pb2.Document()
-        d.id = doc_id
-        d.buffer = text.encode('utf8')
-        d.weight = weight
-        d.length = length
-        return d
-
-    with indexer as idx:
-        data = {
-            'd1': MessageToJson(create_document(1, 'cat', 0.1, 3)),
-            'd2': MessageToJson(create_document(2, 'dog', 0.2, 3)),
-            'd3': MessageToJson(create_document(3, 'bird', 0.3, 3)),
-        }
-        idx.add(data)
+cur_dir = Path(__file__).parent.absolute()
+
+
+def create_document(doc_id, text, weight, length):
+    d = Document()
+    d._document.id = (str(doc_id) * 16)[:16]
+    d.buffer = text.encode('utf8')
+    d.weight = weight
+    d.length = length
+    return d
+
+
+def get_documents(ids):
+    documents = [
+        [0, 'cat', 0.1, 3],
+        [1, 'dog', 0.2, 3],
+        [2, 'crow', 0.3, 4],
+        [3, 'pikachu', 0.4, 7],
+        [4, 'magikarp', 0.5, 8]
+    ]
+    id_to_document = {d[0]: d for d in documents}
+    data = [
+        (id_to_document[id][0], id_to_document[id], MessageToJson(create_document(*id_to_document[id])).encode('utf-8'))
+        for id in ids
+    ]
+    return zip(*data)
+
+
+def apply_action(idx, action):
+    method, key_to_id = action
+    with idx:
+        keys = key_to_id.keys()
+        ids = key_to_id.values()
+        ids, documents, documents_bytes = get_documents(ids)
+        if method == 'add':
+            idx.add(keys, documents_bytes)
+        elif method == 'delete':
+            idx.delete(keys)
+        elif method == 'update':
+            idx.update(keys, documents_bytes)
+        else:
+            print(f'method {method} is not implemented')
         idx.touch()
         idx.save()
-        save_abspath = idx.save_abspath
-        index_abspath = idx.index_abspath
-    assert os.path.exists(index_abspath)
-    assert os.path.exists(save_abspath)
 
-    with BaseIndexer.load(save_abspath) as searcher:
-        doc = searcher.query('d2')
-        assert doc.id == 2
-        assert doc.length == 3
 
-    rm_files([save_abspath, index_abspath])
+def apply_actions(save_abspath, index_abspath, actions):
+    for action in actions:
+        idx = BaseIndexer.load(save_abspath)
+        apply_action(idx, action)
+    return save_abspath, index_abspath
 
 
-def test_add_query():
-    indexer = LevelDBIndexer(level='doc', index_filename='leveldb.db')
-    run_test(indexer)
+def validate_positive_results(keys, documents, searcher):
+    for key, query_doc in zip(keys, documents):
```

Comment:
assert length of `zip(keys, documents)` is larger than 0 to ensure the for is run

Reply:
good idea. check